cloudinit/url_helper.py

```diff
@@ -423,19 +423,14 @@ def readurl(
     if retries:
         manual_tries = max(int(retries) + 1, 1)
 
+    user_agent = "Cloud-Init/%s" % (version.version_string())
     def_headers = {
```

Comment:
if the purpose of the def_headers is to avoid changing the caller's dictionary, perhaps we should just copy it:

```suggestion
    user_agent = "Cloud-Init/%s" % (version.version_string())
    if headers is not None:
        headers = headers.copy()
```

Reply:
that combined with the headers check below covers all the bases